Create Help Command and Refactor

AbnerSquared opened the pull request:
## Related Issue
Fixes #227 

## Refactors
- Updated all dependencies to their latest version.
- Updated all .NET Standard projects to 2.1 (from 2.0).
- Updated all .NET Core projects to 3.1 (from 3.0).
- Set all projects to have their `LangVersion` as `latest`.
- Modified `MiunieUserTypeReader` to adhere to StyleCop standards.
- Modified `EmbedConstructor.FormatReputationType(ReputationType)` to use a `switch statement` instead.
- Modified `EmbedConstructor` to include a new variable called `DefaultEmbedColor`, which now stores the hexadecimal value of the default `Embed` color, which replaces all `.WithColor(R, G, B)` methods used to `.WithColor(DefaultEmbedColor)`.
- Remove unused `using` in `DiscordMessagesAdapter`.

## Changes
- Create `HelpService`, `HelpResult`, `HelpSection`, `HelpStrings.resx`, `CommandInfoExtensions`, `HelpCommand`, and `ExamplesAttribute`.
- Modified `Strings.resx` and `PhraseKey` to include `USER_EMBED_HELP_TITLE`, `HELP_SUMMARY_TITLE`, `HELP_SUMMARY_EMPTY`, `HELP_EXAMPLE_TITLE`, and `HELP_EXAMPLE_EMPTY`.
- Modified `CommandService` to include a new method called `GetHelpService(ILanguageProvider)`, which returns a new `HelpService`.
- Modified `EmbedConstructor` to include a new method called `CreateHelpEmbed(HelpResult)`, which now creates a new `Embed` based off of the `HelpResult` provided.
- Modified `MiscCommands`, `TimeCommand`, `ProfileCommand`, and `RemoteRepositoryCommand` to have all of their summaries set to the specified ID, as well as their modules having an explicitly specified name.

## Expected Feedback
- Determine if there is any methods that might be out of place.
- Determine if all of the summary values and names are valid.

petrspelos commented on src/Miunie.Core/Strings.resx:
```diff
@@ -158,6 +158,22 @@ Source code available at: {1}</value>
     <comment>No formatting options.</comment>
   </data>
+  <data name="HELP_EXAMPLE_EMPTY" xml:space="preserve">
+    <value>No examples were provided.</value>
```
We should probably be more creative and utilize variance with `{{OR}}`.
What about something like:
```
You'll figure it out.{{OR}}
You know what? I've never used this command.{{OR}}
Yeah, you're on your own with this one, buddy.
```

Or similar. 🙂

petrspelos commented on src/Miunie.Core/Strings.resx:
```diff
@@ -158,6 +158,22 @@ Source code available at: {1}</value>
+    <comment>No formatting options.</comment>
+  </data>
+  <data name="HELP_SUMMARY_EMPTY" xml:space="preserve">
+    <value>No summary was provided.</value>
```
Same thing here, we can probably come up with something fun for this.

petrspelos commented on src/Miunie.Discord/CommandModules/HelpCommand.cs:
```diff
@@ -0,0 +1,46 @@
+        }
+
+        [Command("help")]
+        [Summary("HELP_HELP")]
```
I think there is no practical need for a `.resx` of our summaries. You would have to switch to and from the `.resx` file and you can still very easily just misspell it.

Not to mention the summaries are usually fairly short in nature.

We also don't and won't translate the Discord layer at all, so given the commands are for example in plain strings, I think Summaries should be, too.

Also, we should probably get the examples in as well, don't you think? 🤔

AbnerSquared replied:
I was kind of assuming this as well. I'll get on to reverting it.

petrspelos commented on src/Miunie.Discord/CommandModules/RemoteRepositoryCommand.cs:
```diff
@@ -33,7 +34,7 @@ public RemoteRepositoryCommand(RemoteRepositoryService remoteRepoService, Entity
         }
 
         [Command("repo")]
-        [Summary("Shows the official remote repository hosting the code of this bot")]
```
(about the `.resx`)
Especially since we already had a summary written out in a plain string. I'm pretty sure we discussed it, but decided against `.resx`.

petrspelos commented on src/Miunie.Discord/Embeds/EmbedConstructor.cs:
```diff
@@ -26,13 +27,31 @@ namespace Miunie.Discord.Embeds
     internal static class EmbedConstructor
     {
         private static readonly int RepLogPageSize = 10;
+        private static readonly uint DefaultEmbedColor = 0xEC407A; // 236, 64, 122
```
```suggestion
        private static readonly uint DefaultEmbedColor = 0xEC407A;
```

It's not implicitly understandable what color that is by the comment. So might as well not have it. 😅

petrspelos replied:
I mean, I understand it's most likely the Miunie-pink, but I don't think it matters as much to warrant an explanation.

AbnerSquared replied:
Would you want to rename the color to `MiuniePinkColor` instead? It would explicitly specify the kind of color it is.

petrspelos replied:
Sure 😊

petrspelos commented on src/Miunie.Discord/Entities/HelpResult.cs:
```diff
@@ -0,0 +1,26 @@
+    {
+        public string Title { get; set; }
+
+        public List<HelpSection> Sections { get; } = new List<HelpSection>();
```
Instead of using a concrete `List`, we should use one of the abstract variants, for example `IEnumerable` or `ICollection` if `Add` is important.

petrspelos commented on src/Miunie.Discord/Miunie.Discord.csproj:
```diff
@@ -23,4 +24,19 @@
+      <Generator>PublicResXFileCodeGenerator</Generator>
+      <LastGenOutput>HelpStrings.Designer.cs</LastGenOutput>
+    </EmbeddedResource>
+  </ItemGroup>
```
As mentioned before, let's not use the `.resx`.

petrspelos commented on src/Miunie.Discord/Services/HelpService.cs:
```diff
@@ -0,0 +1,129 @@
+
+namespace Miunie.Discord
+{
+    public class HelpService
```
This is not really a `Service`. We would confuse the `Core` Services with these.

I think a better fit would be `CommandHelpProvider`. Also because the word `Help` might be a bit too general. 👍

petrspelos replied:
Also if you called it a `Service` because of  `CommandService`, keep in mind that comes from `Discord.Net` and has nothing to do with hour architecture. I can see how it may look confusing. 😅

AbnerSquared replied:
No, it makes sense. I actually left it as `HelpService`, because I couldn't think of a proper name for it at the time.

petrspelos commented on src/Miunie.Discord/Services/HelpService.cs:
```diff
@@ -0,0 +1,129 @@
+
+        public HelpResult GetDefault()
+        {
+            HelpResult entry = new HelpResult
```
```suggestion
            var entry = new HelpResult
```

Use implicit declaration where obvious.

petrspelos commented on src/Miunie.Discord/Services/HelpService.cs:
```diff
@@ -0,0 +1,129 @@
+                    .Select(x => $"`{x.Key}`"));
+
+                entry.Sections.Add(new HelpSection(title, content));
+            }
```
It would be much better to keep `entry.Sections` as `IEnumerable`, and extract the entire `foreach` into a well named method such as `GetAllModuleHelpSections()`.

That way this whole method would become trivial with:

```cs
return new HelpResult
{
    Title = _lang.GetPhrase(PhraseKey.USER_EMBED_HELP_TITLE.ToString()),
    Sections = GetAllModuleHelpSections()
};
```

> ⚠️ and of course use implicit declaration where possible...

DraxCodes commented:
We really should be targeting a specific LangVersion rather than latest. 
* It's more descriptive as to what we are using
* Using `latest` will result in getting new features before we are are ready to adopt them.

AbnerSquared commented:
Got it. I'll be honest here and say that I made the version updates blindly, and by the time I noticed when submitting a pull request, I left it in if it would fit. I'll get on those reverts, and fix the string entries. :+1: Would you suggest that I explicitly specify the `LangVersion` to `8.0`?

DraxCodes commented:
> Got it. I'll be honest here and say that I made the version updates blindly, and by the time I noticed when submitting a pull request, I left it in if it would fit. I'll get on those reverts, and fix the string entries. +1 Would you suggest that I explicitly specify the `LangVersion` to `8.0`?

Certainly, setting it specifically to 8.0 should be fine.

AbnerSquared commented:
thoughts about a possible extension? this was made to handle the `string.IsNullOrWhiteSpace(value) ? "" : value`, so that it could be simplified into one method.

```cs
internal static string ValueOrDefault(this string value, string fallback = "")
            => string.IsNullOrWhiteSpace(value)
            ? fallback ?? string.Empty
            : value;
```

AbnerSquared commented:
Another suggested idea I had was to implement `CommandHandler.GetPrefix()`, which would return the prefix used for Miunie and allow it to be used within an example, such as `$"{_prefix}{example}"`.

petrspelos commented:
I think these are great suggestions, @AbnerSquared. And I think it would be great to get your current issue merged in, and create new issues for the prefix for example.

It can also be something for beginners if it's simple enough. 😊

AbnerSquared commented:
This update should fix all issues previously mentioned and make some methods easier to handle.

Changes
- Created `StringExtensions`, which contains two methods of `ValueOrDefault(this string, string)` and `JoinOrDefault(this IEnumerable<string>, string, string)`.
- Created `TExtensions`, which contains one method of `StringJoinOrDefault(this IEnumerable<T>, Func<T, string>, string, string)`.
- Set all `LangVersion` properties to `8.0`.
- Modified `HelpCommand`, `MiscCommands`, `ProfileCommand`, `RemoteRepositoryCommand`, and `TimeCommand` to store their direct summaries and include examples.
- Modified HelpCommandProvider to include three new methods of `GetAllModuleSections()`, `GetSection(ModuleInfo)`, and `GetSection(CommandInfo)`.
- Modified `HelpCommandProvider.Search` to use `GetSection(Command)`.
- Modified `HelpCommandProvider.GetDefault` to use `GetAllModuleSections()`.
- Modified `HelpCommandProvider.GetExamples(CommandInfo)` to use `.StringJoinOrDefault()`.
- Modified `HelpResult.Sections` to implement `IEnumerable<HelpSection>` instead.
- Modified `HelpResult.Sections` to expose the `set` property.
- Modified `Strings.resx` to make `HELP_SUMMARY_EMPTY` and `HELP_EXAMPLE_EMPTY` include multiple response values.
- Modified `EmbedConstructor.GetHelpEmbed(HelpResult)` to use `MiuniePinkColor` for its `EmbedBuilder`.
- Renamed `EmbedConstructor.DefaultEmbedColor` to `MiuniePinkColor` and removed comment.
- Renamed `HelpService` to `HelpCommandProvider`.
- Renamed `CommandService.GetHelpService(ILanguageProvider)` to `CommandService.GetHelpProvider(ILanguageProvider)`.
- Reverted all .NET Standard targets back to 2.0.
- Deleted `HelpStrings.resx`.

AbnerSquared commented:
## Patches
- Modified `StringExtensions.JoinOrDefault()` to use the proper method when determining the collections count.

AbnerSquared commented:
## Formatting
- Modified `EmbedConstructor.GetHelpEmbed(HelpResult)` to use implicit declaration.
- Modified `HelpCommandProvider.GetAllModuleSections` to include a blank space in between a variable and a `foreach` statement.

petrspelos commented on src/Miunie.Core/Strings.resx:
```diff
@@ -158,6 +158,28 @@ Source code available at: {1}</value>
+You know what? I've never used this command.{{OR}}
+Yeah, you're on your own with this one, buddy.{{OR}}
+I got nothing here.{{OR}}
+Just mess around with this, you'll figure it out eventually.</value>
```
I love this one. 😂

petrspelos commented on src/Miunie.Discord/CommandModules/HelpCommand.cs:
```diff
@@ -0,0 +1,49 @@
+    [Name("Help")]
+    public class HelpCommand : ModuleBase<SocketCommandContext>
+    {
+        private readonly HelpCommandProvider _helpService;
```
This is still called a `_helpService`.

petrspelos commented on src/Miunie.Discord/CommandModules/ProfileCommand.cs:
```diff
@@ -62,6 +70,8 @@ public async Task ShowReputationLogAsync(MiunieUser user, int page = 1)
         }
 
         [Command("+rep")]
+        [Summary("Give a good ol' positive reputation for a user!")]
```
```suggestion
        [Summary("Give a good ol' positive reputation to a user!")]
```

petrspelos commented on src/Miunie.Discord/CommandModules/ProfileCommand.cs:
```diff
@@ -70,6 +80,8 @@ public async Task AddReputationAsync(MiunieUser user)
         }
 
         [Command("-rep")]
+        [Summary("Give a cold-blooded negative reputation for a user...")]
```
```suggestion
        [Summary("Give a cold-blooded negative reputation to a user...")]
```

petrspelos commented on src/Miunie.Discord/Providers/HelpCommandProvider.cs:
```diff
@@ -0,0 +1,129 @@
+                sections.Add(GetSection(module));
+            }
+
+            return sections;
```
I think this can just be:

```cs
return _commandService.Modules.Select(GetSection);
```

instead of the entire method. And honesty, at that point we don't even need this method to exist.

We could just use the suggested one-liner where `GetAllModuleSelections` is used. 🙂

petrspelos commented on src/Miunie.Discord/Providers/HelpCommandProvider.cs:
```diff
@@ -0,0 +1,129 @@
+        private HelpSection GetSection(ModuleInfo module)
+        => new HelpSection(module.Name, string.Join(" ", module.Commands
+            .GroupBy(x => x.Name)
+            .Select(x => $"`{x.Key}`")));
```
The formatting of this method isn't very clear.

You are trying to one-line waaaaay too much.

Potential fix could be a method extraction refactoring or maybe even just having a body.

AbnerSquared commented:
I fixed all issues, but would you want to preserve the current `HelpCommandProvider.Search(string input)` method the way it is for clarity? I wrote another version that implements the same `.Select()` method that `HelpCommandProvider.GetDefault()` now uses.

```cs
public HelpResult Search(string input)
        {
            var sections = new List<HelpSection>();
            foreach (CommandInfo command in GetCommands(input))
            {
                sections.Add(GetSection(command));
            }

            return new HelpResult()
            {
                Sections = sections
            };
        }
```

```cs
public HelpResult Search(string input)
        => new HelpResult()
        {
            Sections = GetCommands(input).Select(x => GetSection(x))
        };
```

AbnerSquared commented:
## Changes
- Modified `HelpCommandProvider` to include a new method of `GetModuleCommandBlocks(ModuleInfo)`.
- Modified `ProfileCommand` to use "to" instead of "for" in both summaries.
- Modified `HelpCommandProvider.GetDefault()` to use `_commandService.Modules.Select()` instead.
- Modified `HelpCommandProvider.GetSection(ModuleInfo)` to use `GetModuleCommandBlocks()`.
- Removed `HelpCommandProvider.GetAllModuleHelpSections()`.

AbnerSquared commented:
## Changes
- Modified `HelpCommandProvider.Search(string)` to use the similar `.Select()` method used in `GetDefault()`.

petrspelos commented on src/Miunie.Discord/Extensions/TExtensions.cs:
```diff
@@ -0,0 +1,33 @@
+            => values?.Count() > 0
+            ? string.Join(separator, values.Select(x => formatter.Invoke(x)))
+            : fallback ?? string.Empty;
+    }
```
This is only referenced once in:

```cs
command.FindAttribute<ExamplesAttribute>()?.Examples
                .StringJoinOrDefault(x => $"`{x}`", ", ", _lang.GetPhrase(PhraseKey.HELP_EXAMPLE_EMPTY.ToString()));
```

which could very well be something like

```cs
var examples = command.FindAttribute<ExamplesAttribute>()?.Examples;

if (!examples.Any()) { return _lang.GetPhrase(PhraseKey.HELP_EXAMPLE_EMPTY.ToString()); }

return string.Join(", ", examples.Select(x => $"`{x}`"));
```

Which is not nearly as complex to understand and does the job.

YAGNI, that was the only reference of this method, so I'd say let's nuke them. 😊

AbnerSquared commented on src/Miunie.Discord/CommandHandler.cs:
```diff
@@ -53,6 +54,9 @@ public async Task InitializeAsync()
 
+        public HelpCommandProvider GetHelpProvider(ILanguageProvider lang)
+            => new HelpCommandProvider(_commandService, lang);
+
```
Would it be fine to initialize the `CommandService` in the IoC? The reason I made this method was because the `CommandService` was initialized in `CommandHandler`.

petrspelos replied:
The `CommandService` should definitely be injected as well.
I think there was some issue with it not resolving properly, but we need it to be properly injected.

AbnerSquared commented on src/Miunie.Discord/Extensions/StringExtensions.cs:
```diff
@@ -0,0 +1,36 @@
+            => values?.Count() > 0
+            ? string.Join(separator, values)
+            : fallback ?? string.Empty;
+    }
```
Yeah, my bad. It was originally for a specific command and forgot to remove it when implementing `StringJoinOrDefault`

AbnerSquared commented on src/Miunie.Discord/Providers/HelpCommandProvider.cs:
```diff
@@ -0,0 +1,112 @@
+        private string GetSectionParameters(CommandInfo command)
+            => string.Join(" ", command.Parameters
+                .OrderBy(x => x.IsOptional)
+                .Select(x => x.IsOptional ? $"[{x.Name}]" : $"<{x.Name}>"));
```
Yeah, that makes sense to follow a pattern of writing variables with only a single layer of depth.

AbnerSquared commented on src/Miunie.Discord/Providers/HelpCommandProvider.cs:
```diff
@@ -0,0 +1,112 @@
+
+namespace Miunie.Discord
+{
+    public class HelpCommandProvider
```
Actually, I think you mentioned to use `CommandHelpProvider` beforehand. That was my bad.

AbnerSquared replied:
Honestly, I think I depend a bit too much on summaries and comments for my code. While I will still use them, I'll try to be more explicit when naming.

AbnerSquared commented:
I'll be able to make these changes in around an hour.

AbnerSquared commented:
## Changes
- Modify constructor `HelpCommand(CommandHandler, ILanguageProvider)` to instead be initialized as `HelpCommand(CommandService, ILanguageProvider)`
- Modify method `AddMiunieTypes(this IServiceCollection)` in `Container` to include `CommandService` as a singleton
- Modify method `GetHelp()` in `HelpCommand` to handle sending the message result
- Modify method `GetHelp(string)` in `HelpCommand` to handle sending the message result
- Renamed `HelpCommandProvider` to `CommandHelpProvider`
- Rename method `GetDefault()` to `ForAllCommands()` in `CommandHelpProvider`
- Rename method `Search(string)` to `FromInput(string)` in `CommandHelpProvider`
- Remove method `GetHelpProvider(ILanguageProvider)` from `CommandHandler`
- Remove method `ShowDefaultHelpAsync(IMessageChannel)` in `CommandHelpProvider`
- Remove method `ShowCommandHelpAsync(IMessageChannel, string)` from `CommandHelpProvider`
- Delete `StringExtensions`
- Delete `TExtensions`

AbnerSquared commented:
@petrspelos 

There is currently only one concern I have before I think I'm good, and that is if the constructor for `HelpCommand` is viable.

```cs
public HelpCommand(CommandService commandService, ILanguageProvider lang)
        {
            _helpProvider = new CommandHelpProvider(commandService, lang);
        }
```

This is its current state, but for me personally, it doesn't look like it follows the rule set that is wanted when initializing new classes from DI, and I just wanted to make sure before leaving it as is.

petrspelos commented on src/Miunie.Discord/Providers/CommandHelpProvider.cs:
```diff
@@ -0,0 +1,129 @@
+        {
+            var examples = command.FindAttribute<ExamplesAttribute>()?.Examples;
+
+            if (!examples?.Any() ?? true)
```
```suggestion
            if (examples is null || !examples.Any())
```
No need to overcomplicate things.

petrspelos commented:
Tested on `elementary OS 5.1.2 Hera x86_64` GNU/Linux.
Front-end: `Miunie.ConsoleApp`
Commit: add39db
Result: :heavy_check_mark: Feature works as expected.

Awesome job, @AbnerSquared! :tada:
